Insight Password verification failed

<!-- Thanks for filing an issue! Before hitting the button, please answer these questions. 
It's helpful to search the existing GitHub issues first. It's likely that another user 
has already reported the issue you're facing, or it's a known issue that we're already aware of-->

**Is this a request for help?**:
Yes
---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one):
BUG REPORT
<!--
If this is a BUG REPORT, please:
  - Fill in as much of the template below as you can.  If you leave out
    information, we can't help you as well.

If this is a FEATURE REQUEST, please:
  - Describe *in detail* the feature/behavior/change you'd like to see.

In both cases, be ready for followup questions, and please respond in a timely
manner.  If we can't reproduce a bug or think a feature already exists, we
might close your issue.  If we're wrong, PLEASE feel free to reopen it and
explain why.
-->

**Version of Helm and Kubernetes**:
helm version
Client: &version.Version{SemVer:"v2.9.1", GitCommit:"20adb27c7c5868466912eebdf6664e7390ebe710", GitTreeState:"clean"}
Server: &version.Version{SemVer:"v2.9.1", GitCommit:"20adb27c7c5868466912eebdf6664e7390ebe710", GitTreeState:"clean"}

kubectl version
Client Version: version.Info{Major:"1", Minor:"9", GitVersion:"v1.9.7", GitCommit:"dd5e1a2978fd0b97d9b78e1564398aeea7e7fe92", GitTreeState:"clean", BuildDate:"2018-04-19T00:05:56Z", GoVersion:"go1.9.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.7-gke.6", GitCommit:"9b635efce81582e1da13b35a7aa539c0ccb32987", GitTreeState:"clean", BuildDate:"2018-08-16T21:33:47Z", GoVersion:"go1.9.3b4", Compiler:"gc", Platform:"linux/amd64"}

**Which chart**:
mission-control version 0.4.4

**What happened**:
The following error below is encountered upon deployment of mission control helm chart even when generate_keys.sh was executed and following commands were executed to create and pass secret to deployment:
kubectl create secret generic mission-control-certs --from-file=./certs/insight-server/etc/security/insight.key --from-file=./certs/insight-server/etc/security/insight.crt --from-file=./certs/insight-server/etc/security/jfmc.crt  --from-file=./certs/mission-control/etc/security/jfmc-truststore.jks-b64 --from-file=./certs/mission-control/etc/security/jfmc-keystore.jks-b64

helm install --name mission-control --set existingCertsSecret=mission-control-certs jfrog/mission-control

```
2018-10-04 21:46:54,596 [pool-6-thread-2] [ERROR] (o.j.m.s.i.InsightService:59) - Error occurred while refreshing Insight services.
org.jfrog.mc.service.exception.ServiceException: Error occured while creating the SSLContext: Keystore was tampered with, or password was incorrect
	at org.jfrog.mc.insight.InsightClient.createSslContext(InsightClient.java:213)
	at org.jfrog.mc.insight.InsightClient.createSslClient(InsightClient.java:135)
	at org.jfrog.mc.insight.InsightClient.doSslExecute(InsightClient.java:165)
	at org.jfrog.mc.insight.InsightClient.replaceServices(InsightClient.java:100)
	at org.jfrog.mc.service.insight.InsightService.replaceServices(InsightService.java:53)
	at org.jfrog.mc.service.RefreshInstanceDataServiceImpl.replaceAllInstancesOnInsight(RefreshInstanceDataServiceImpl.java:103)
	at org.jfrog.mc.service.RefreshInstanceDataServiceImpl.replaceAllInstancesOnInsight(RefreshInstanceDataServiceImpl.java:97)
	at org.jfrog.mc.service.RefreshInstanceDataServiceImpl$$FastClassBySpringCGLIB$$26fc11a2.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:738)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:673)
	at org.jfrog.mc.service.RefreshInstanceDataServiceImpl$$EnhancerBySpringCGLIB$$87aa8adf.replaceAllInstancesOnInsight(<generated>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:333)
	at org.springframework.aop.framework.JdkDynamicAopProxy.invoke(JdkDynamicAopProxy.java:207)
	at com.sun.proxy.$Proxy164.replaceAllInstancesOnInsight(Unknown Source)
	at org.jfrog.mc.service.AsyncRefreshInstanceDataService.replaceAllInstancesOnInsight(AsyncRefreshInstanceDataService.java:52)
	at org.jfrog.mc.service.AsyncRefreshInstanceDataService$$FastClassBySpringCGLIB$$7dfa1990.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:204)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.invokeJoinpoint(CglibAopProxy.java:738)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:157)
	at org.springframework.aop.framework.adapter.MethodBeforeAdviceInterceptor.invoke(MethodBeforeAdviceInterceptor.java:52)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:168)
	at org.springframework.aop.interceptor.ExposeInvocationInterceptor.invoke(ExposeInvocationInterceptor.java:92)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:179)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:673)
	at org.jfrog.mc.service.AsyncRefreshInstanceDataService$$EnhancerBySpringCGLIB$$3ba341ea.replaceAllInstancesOnInsight(<generated>)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.springframework.scheduling.support.ScheduledMethodRunnable.run(ScheduledMethodRunnable.java:65)
	at org.springframework.scheduling.support.DelegatingErrorHandlingRunnable.run(DelegatingErrorHandlingRunnable.java:54)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180)
	at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Keystore was tampered with, or password was incorrect
	at sun.security.provider.JavaKeyStore.engineLoad(JavaKeyStore.java:780)
	at sun.security.provider.JavaKeyStore$JKS.engineLoad(JavaKeyStore.java:56)
	at sun.security.provider.KeyStoreDelegator.engineLoad(KeyStoreDelegator.java:224)
	at sun.security.provider.JavaKeyStore$DualFormatJKS.engineLoad(JavaKeyStore.java:70)
	at java.security.KeyStore.load(KeyStore.java:1445)
	at org.jfrog.mc.insight.InsightClient.createSslContext(InsightClient.java:196)
	... 45 common frames omitted
Caused by: java.security.UnrecoverableKeyException: Password verification failed
	at sun.security.provider.JavaKeyStore.engineLoad(JavaKeyStore.java:778)
	... 50 common frames omitted
```

**What you expected to happen**:
Keystore should have been set properly for Insight and error should not occur.

**How to reproduce it** (as minimally and precisely as possible):
1. Create GKE cluster
2. Clone helm charts 
3. Execute generate_keys.sh
4. Modify values.yaml with the minimum requirements 
5. Generate the secret
6. Deploy mission-control helm chart with created secret
7. Microservices will be deployed and running but Mission Control log will have the error above.

**Anything else we need to know**:

--- a/stable/mission-control/CHANGELOG.md
+++ b/stable/mission-control/CHANGELOG.md
@@ -1,6 +1,9 @@
 # JFrog Mission-Control Chart Changelog
 All notable changes to this chart will be documented in this file.
 
+## [0.4.6] - Oct 16, 2018
+* Fix #67 Set password used to generate internal certs in Mission-Control
+
 ## [0.4.5] - Oct 9, 2018
 * Quote ingress hosts to support wildcard names
 

--- a/stable/mission-control/Chart.yaml
+++ b/stable/mission-control/Chart.yaml
@@ -1,7 +1,7 @@
 apiVersion: v1
 name: mission-control
 description: A Helm chart for JFrog Mission Control
-version: 0.4.5
+version: 0.4.6
 appVersion: 3.1.2
 home: https://jfrog.com/mission-control/
 icon: https://raw.githubusercontent.com/JFrogDev/artifactory-dcos/master/images/jfrog_med.png

--- a/stable/mission-control/README.md
+++ b/stable/mission-control/README.md
@@ -165,6 +165,7 @@ arrangeFiles() {
     mv -f ${tmpDir}/insight.crt ${insightSecurity}
     cat ${jfmcSecurity}/jfmc-truststore.jks | base64 > ${jfmcSecurity}/jfmc-truststore.jks-b64
     cat ${jfmcSecurity}/jfmc-keystore.jks | base64 > ${jfmcSecurity}/jfmc-keystore.jks-b64
+    echo "${storePassword}" | base64 > ${jfmcSecurity}/jfmc-store-passwd-b64
 }
 
 summary() {

--- a/stable/mission-control/ci/test-values.yaml
+++ b/stable/mission-control/ci/test-values.yaml
@@ -165,3 +165,5 @@ jfmcTruststore: |
   FnbvhQrLzB05CeCLAgtrYBO2HgVvWC7ecv9uT1q7/Sj02YyEmGfsx+A8ic/0QUu7
   4ci08xUjLq9zcCuLh5rW8eK+sH2k1zbUqrmdMknpfUSVe+u3aMiixUWQQe3t33Zb
   c4e6ekZju53Q39bQ4H9OgJU=
+
+jfmcStorePasswd : MThmODVjMzMxZjVlM2NkNAo=

--- a/stable/mission-control/templates/mission-control-certs.yaml
+++ b/stable/mission-control/templates/mission-control-certs.yaml
@@ -19,4 +19,6 @@ data:
   jfmc-keystore.jks-b64: {{ required "A valid .Values.jfmcKeystore entry required!" .Values.jfmcKeystore | b64enc | quote }}
 
   jfmc-truststore.jks-b64: {{ required "A valid .Values.jfmcTruststore entry required!" .Values.jfmcTruststore | b64enc | quote }}
+
+  jfmc-store-passwd-b64: {{ required "A valid .Values.jfmcStorePasswd entry required!" .Values.jfmcStorePasswd | b64enc | quote }}
 {{- end }}

--- a/stable/mission-control/templates/mission-control-deployment.yaml
+++ b/stable/mission-control/templates/mission-control-deployment.yaml
@@ -38,6 +38,44 @@ spec:
                 nc -z -w 2 {{ .Release.Name }}-elasticsearch 9200 && echo elasticsearch ok;
             do sleep 2;
           done
+      - name: set-password
+        image: "{{ .Values.initContainerImage }}"
+        command:
+        - 'sh'
+        - '-c'
+        - >
+          until [ -f /tmp/jfmc-keystore.jks-b64 ] && [ -f /tmp/jfmc-truststore.jks-b64 ]; do sleep 1; done;
+          mkdir -p {{ .Values.missionControl.persistence.mountPath }}/etc/security {{ .Values.missionControl.persistence.mountPath }}/logs {{ .Values.missionControl.persistence.mountPath }}/data/security;
+          base64 -d /tmp/jfmc-keystore.jks-b64 > {{ .Values.missionControl.persistence.mountPath }}/etc/security/jfmc-keystore.jks;
+          base64 -d /tmp/jfmc-truststore.jks-b64 > {{ .Values.missionControl.persistence.mountPath }}/etc/security/jfmc-truststore.jks;
+
+          if [ -f /tmp/jfmc-store-passwd-b64 ]; then                                                          \
+            propertyName=server.internal.ssl.trust-store-password;                                            \
+            propertiesPath={{ .Values.missionControl.persistence.mountPath }}/etc/mission-control.properties; \
+            propertyValue=$(base64 -d /tmp/jfmc-store-passwd-b64);                                            \
+                                                                                                              \
+            grep "^${propertyName}\s*=.*$" ${propertiesPath} > /dev/null 2>&1;                                \
+            [ $? -ne 0 ] && echo "${propertyName}=${propertyValue}" >> ${propertiesPath};                     \
+            sed -i -e "s|^${propertyName}\s*=.*$|${propertyName}=${propertyValue}|g;" ${propertiesPath};      \
+                                                                                                              \
+            propertyName=server.internal.ssl.key-store-password;                                              \
+            grep "^${propertyName}\s*=.*$" ${propertiesPath} > /dev/null 2>&1;                                \
+            [ $? -ne 0 ] && echo "${propertyName}=${propertyValue}" >> ${propertiesPath};                     \
+            sed -i -e "s|^${propertyName}\s*=.*$|${propertyName}=${propertyValue}|g;" ${propertiesPath};      \
+          fi;
+          chown -R {{ .Values.uid }}:{{ .Values.uid }} {{ .Values.missionControl.persistence.mountPath }};
+        volumeMounts:
+        - name: mission-control-data
+          mountPath: {{ .Values.missionControl.persistence.mountPath | quote }}
+        - name: mission-control-certs
+          mountPath: /tmp/jfmc-keystore.jks-b64
+          subPath: jfmc-keystore.jks-b64
+        - name: mission-control-certs
+          mountPath: /tmp/jfmc-truststore.jks-b64
+          subPath: jfmc-truststore.jks-b64
+        - name: mission-control-certs
+          mountPath: /tmp/jfmc-store-passwd-b64
+          subPath: jfmc-store-passwd-b64
       containers:
       - name: {{ .Values.missionControl.name }}
         image: {{ .Values.missionControl.image }}:{{ default .Chart.AppVersion .Values.missionControl.version }}
@@ -60,10 +98,6 @@ spec:
           value: "https://{{ template "insight-server.fullname" . }}:{{ .Values.insightServer.internalHttpsPort }}"
         - name: POD_RESTART_TIME
           value: "{{ .Values.podRestartTime }}"
-        - name: SERVER_INTERNAL_SSL_KEY_STORE_PASSWORD
-          value: "18f85c331f5e3cd4"
-        - name: SERVER_INTERNAL_SSL_TRUST_STORE_PASSWORD
-          value: "18f85c331f5e3cd4"
         - name: ARTIFACTORY_CLIENT_CONNECTIONTIMEOUT
           value: '20'
         - name: XRAY_CLIENT_CONNECTIONTIMEOUT
@@ -94,23 +128,6 @@ spec:
         volumeMounts:
         - name: mission-control-data
           mountPath: {{ .Values.missionControl.persistence.mountPath | quote }}
-        - name: mission-control-certs
-          mountPath: /tmp/jfmc-keystore.jks-b64
-          subPath: jfmc-keystore.jks-b64
-        - name: mission-control-certs
-          mountPath: /tmp/jfmc-truststore.jks-b64
-          subPath: jfmc-truststore.jks-b64
-        lifecycle:
-          postStart:
-            exec:
-              command:
-              - '/bin/sh'
-              - '-c'
-              - >
-                until [ -f /tmp/jfmc-keystore.jks-b64 ] && [ -f /tmp/jfmc-truststore.jks-b64 ]; do sleep 1; done;
-                mkdir -p /var/opt/jfrog/mission-control/etc/security;
-                base64 -d /tmp/jfmc-keystore.jks-b64 > /var/opt/jfrog/mission-control/etc/security/jfmc-keystore.jks;
-                base64 -d /tmp/jfmc-truststore.jks-b64 > /var/opt/jfrog/mission-control/etc/security/jfmc-truststore.jks
         resources:
 {{ toYaml .Values.missionControl.resources | indent 10 }}
         livenessProbe:

--- a/stable/mission-control/values.yaml
+++ b/stable/mission-control/values.yaml
@@ -7,6 +7,8 @@
 initContainerImage: "alpine:3.6"
 imagePullPolicy: IfNotPresent
 
+uid: 1050
+
 imagePullSecrets:
 
 ## Role Based Access Control
@@ -125,6 +127,7 @@ insightCrt:
 jfmcCrt:
 jfmcKeystore:
 jfmcTruststore:
+jfmcStorePasswd:
 
 missionControl:
   replicaCount: 1

